internal/pkg/cli/deploy/env.go

```diff
@@ -142,24 +165,82 @@ func NewEnvDeployer(in *NewEnvDeployerInput) (*envDeployer, error) {
 		newServiceStackDescriber: func(svc string) stackDescriber {
 			return stack.NewStackDescriber(cfnstack.NameForService(in.App.Name, in.Env.Name, svc), envManagerSession)
 		},
-	}, nil
+
+		ws: in.Workspace,
+	}
+	deployer.parseAddons = func() (stackBuilder, error) {
+		if !in.EnvAddonsFeatureFlag {
+			return nil, &addon.ErrAddonsNotFound{}
+		}
+		deployer.parseAddonsOnce.Do(func() {
+			deployer.addons.stack, deployer.addons.err = addon.ParseFromEnv(deployer.ws)
+		})
+		return deployer.addons.stack, deployer.addons.err
+	}
+	return deployer, nil
 }
 
 // Validate returns an error if the environment manifest is incompatible with services and application configurations.
 func (d *envDeployer) Validate(mft *manifest.Environment) error {
 	return d.validateCDN(mft)
 }
 
+// UploadEnvArtifactsOutput is the output of UploadArtifacts.
```

Comment:
nit: 😂 maybe to be a bit less redundant:
```suggestion
// UploadEnvArtifactsOutput holds URLs of artifacts pushed to S3 buckets, ECR repositories...
```